### PR TITLE
[SEC-1249] Add SHA256 digest variant for iOS13 + CryptoKit

### PR DESCRIPTION
**Pull request by Simon-F-AMS**
- Add digest variant 2 to SHA256 which uses CryptoKit
- Add additional unit test cases for new variants

I'm not sure about the function names yet so suggestions are welcome 👍

### CODE REVIEW

File: Example/Tests/SHA256Tests.swift [hunk 1]
```diff
@@ -41,4 +41,36 @@ class SHA256Tests: XCTestCase {
     // then
     XCTAssertEqual(bufferedDigest, Data(hex: expectedHash))
   }
+
+  @available(iOS 13, *)
+  func testDigestData_iOS13__shouldProduceExpectedHash() {
```
Comment (hngry) [inline Example/Tests/SHA256Tests.swift]:
```suggestion
  func testDigestData_usingCryptoKit__shouldProduceExpectedHash() {
```

File: Example/Tests/SHA256Tests.swift [hunk 2]
```diff
@@ -41,4 +41,36 @@ class SHA256Tests: XCTestCase {
...
+    // given
+    let testData = TestData.largeTestData
+    // when
+    let digest = SHA256.digestV2(testData.data)
+    // then
+    XCTAssertEqual(digest, Data(hex: expectedHash))
+  }
+
+  @available(iOS 13, *)
+  func testDigestFile_iOS13_whenDataLargeEnoughFor5BufferFills__shouldProduceExpectedHash() {
```
Comment (hngry) [inline Example/Tests/SHA256Tests.swift]:
```suggestion
  func testDigestFile_usingCryptoKit_whenDataLargeEnoughFor5BufferFills__shouldProduceExpectedHash() {
```

File: Example/Tests/SHA256Tests.swift [hunk 3]
```diff
@@ -41,4 +41,36 @@ class SHA256Tests: XCTestCase {
...
+    /// Test data large enough to force at least 5 buffer fills
+    let testData = TestData.largeTestData
+    // when
+    let bufferedDigest = try! SHA256.digestV2(file: testData.url)
+    // then
+    XCTAssertEqual(bufferedDigest, Data(hex: expectedHash))
+  }
+
+  @available(iOS 13, *)
+  func testDigestFile_iOS13_withCustomBufferSize__shouldProduceExpectedHash() {
```
Comment (hngry) [inline Example/Tests/SHA256Tests.swift]:
```suggestion
  func testDigestFile_usingCryptoKit_withCustomBufferSize__shouldProduceExpectedHash() {
```

File: Krypt/Source/SHA256.swift [hunk 4]
```diff
@@ -21,6 +22,13 @@ public struct SHA256 {
     return status == errSecSuccess ? Data(hash) : nil
   }
 
+  /// Calculates SHA256 hash of given data using CryptoKit
+  /// - Parameter data: The data to calculate the hash for
+  @available(iOS 13, *)
+  public static func digestV2(_ data: Data) -> Data? {
```
Comment (hngry) [inline Krypt/Source/SHA256.swift]: There's no v2, should be part of the `digest(_ data: Data) -> Data?` using either CommonCrypto or CryptoKit depending on iOS version.

File: Krypt/Source/SHA256.swift [hunk 5]
```diff
@@ -61,4 +69,34 @@ public struct SHA256 {
 
     return digest
   }
+
+  /// iOS 13 Variant of SHA256 hash calculation for large files using a buffer and CryptoKit
+  /// - Parameters:
+  ///   - url: The url to the file to calculate the SHA256 hash for
+  ///   - withBufferSize: The size of the buffer to use in bytes, defaults to 1024*1024 bytes = 1 MB
+  @available(iOS 13, *)
+  public static func digestV2(file url: URL, withBufferSize: Int = 1024 * 1024) throws -> Data {
```
Comment (hngry) [inline Krypt/Source/SHA256.swift]: same

File: Krypt/Source/SHA256.swift [hunk 6]
```diff
@@ -61,4 +69,34 @@ public struct SHA256 {
...
+      throw SHA256Error.fileOperationError
+    }
+    stream.open()
+    let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: withBufferSize)
+    defer {
+      buffer.deallocate()
+    }
+    var hasher = CryptoKit.SHA256()
+    while stream.hasBytesAvailable {
+      let read = stream.read(buffer, maxLength: withBufferSize)
```
Comment (hngry) [inline Krypt/Source/SHA256.swift]: Also this one still haunts me, you changed it from `with bufferSize: Int` to `withBufferSize: Int` while I was suggesting `withBufferSize bufferSize: Int` which gives you the best of both.
```suggestion
      let read = stream.read(buffer, maxLength: bufferSize)
```

Comment (Simon-F-AMS) [reply Krypt/Source/SHA256.swift]: my bad will fix it

**Review by markotect — APPROVED** [review]
Nice one!

**Pull request merged**